### PR TITLE
Implement BinaryMarshaler for key.Key256 and tighten up zero value behaviour

### PR DESCRIPTION
**Pull request by iand**

### CODE REVIEW

File: key/key.go [hunk 1]
```diff
@@ -86,7 +90,15 @@ func (k Key256) CommonPrefixLength(o Key256) int {
 
 // Compare compares the numeric value of the key with another key of the same type.
 func (k Key256) Compare(o Key256) int {
-	return bytes.Compare(k.b[:], o.b[:])
+	if k.b != nil && o.b != nil {
```
Comment (dennis-tra) [inline key/key.go:93]: could it actually happen that a user of Key256 ends up with a nil `.b` field? I'm just wondering and all good with being defensive here.

File: key/key.go [hunk 2]
```diff
@@ -97,6 +109,16 @@ func (k Key256) HexString() string {
 	return hex.EncodeToString(k.b[:])
 }
 
+// MarshalBinary marshals the key into a byte slice.
+// The bytes may be passed to NewKey256 to construct a new key with the same value.
+func (k Key256) MarshalBinary() ([]byte, error) {
```
Comment (dennis-tra) [inline key/key.go:114]: You could consider an interface assertion along `var _ encoding.BinaryMarshaler = Key256{}` and `var _ encoding.BinaryUnmarshaler = Key256{}`

File: key/key_test.go [hunk 3]
```diff
@@ -230,6 +256,21 @@ func (kt *KeyTester[K]) TestHexString(t *testing.T) {
 	}
 }
 
+// testBinaryMarshaler tests the behaviour of a kad.Key implementation that also implements the BinaryMarshaler interface
+func testBinaryMarshaler[K interface {
+	kad.Key[K]
+	MarshalBinary() ([]byte, error)
+}](t *testing.T, k K, newFunc func([]byte) K,
```
Comment (dennis-tra) [inline key/key_test.go:263]:
```suggestion
func testBinaryMarshaler[K interface {
	kad.Key[K]
	encoding.BinaryMarshaler
}](t *testing.T, k K, newFunc func([]byte) K,
```

Would this work?

**Review by dennis-tra — APPROVED** [review]

File: key/key.go (hunk 1 above)
Comment (iand) [reply key/key.go:93]: `var k key.Key256` will do it

File: key/key.go (hunk 2 above)
Comment (iand) [reply key/key.go:114]: Then I would have to import the encoding package simply for a test assertion. Could do it in key_test.go

**Review by guillaumemichel — APPROVED** [review]

**Pull request merged**